undertow/src/main/java/org/wildfly/extension/undertow/deployment/WarMetaDataProcessor.java

```diff
@@ -249,7 +238,7 @@ public void deploy(DeploymentPhaseContext phaseContext) throws DeploymentUnitPro
         WebCommonMetaData mergedFragmentMetaData = new WebCommonMetaData();
         if (specMetaData == null) {
             // If there is no web.xml, it has to be considered to be the latest version
-            specMetaData = new Web31MetaData();
+            specMetaData = new WebMetaData();
             specMetaData.setVersion("3.1");
```

Comment:
Shall we still stick with 3.1only? No condition based on ee8.preview.mode property value?

Reply:
I have updated this, and also added schema validation for 4.0. There is an additional change that has to be made here to make sure the correct version is reported in TCK tests, but it can't happen till Undertow 2.0, which can't happen till these changes are merged.